### DISCUSSION TITLE
在nuxt3中，使用alova遇到的异步问题

### DISCUSSION THREAD

**Discussion by FatNerdPeng** [alovajs/alova#189] [General]
代码是这样：

![image](https://github.com/alovajs/alova/assets/33616855/60aa0131-b801-4a0c-bd19-8bf0ea390a4c)

“账号”这个表单项，我写了个验证函数：

![image](https://github.com/alovajs/alova/assets/33616855/709132cf-fc3c-45b0-b109-c8b9bb6f87c2)

因为需要，我自己在nuxt的composables目录下，创建了一个封装函数：

![image](https://github.com/alovajs/alova/assets/33616855/28d3ceb6-e0e7-4a12-a7da-f2f97833ad85)
![image](https://github.com/alovajs/alova/assets/33616855/8625b0bf-3674-49b3-b665-d9728c862e39)
![image](https://github.com/alovajs/alova/assets/33616855/1570a3d8-97b5-4991-87f5-6c15f3e47973)

在这个过程中，我写了2个console，我预想的是先输出json，再输出res，但结果是相反的：

![image](https://github.com/alovajs/alova/assets/33616855/54c99c44-cc2f-4f64-be74-77e498bfa1e4)

按照我的猜想，出现这个现象，是41行代码的httpRequest是异步的，所以实际还未发出请求就return了，所以先输出的是res，当然res的各属性都是undefined，然后输出json，返回了服务端响应的内容"{code:1, message:账号不存在}"，我希望不希望是这种不合理的顺序，所以我给useRequest前加上await，但是它提示：

![image](https://github.com/alovajs/alova/assets/33616855/5d71ca25-4acd-4aa3-b916-81991e6ef196)

于是我又换了个思路。按照alova文档所说，nuxt3中通过alova获取同步数据，可以结合nuxt3提供的组合api `useAsyncData` 实现。于是我这样写：

![image](https://github.com/alovajs/alova/assets/33616855/8f1aad52-9dd1-4b8c-bbb5-311994855832)

这次的结果是顺序对了，但是服务端的响应内容前端没有正常收到：

![image](https://github.com/alovajs/alova/assets/33616855/5f46a6b6-d933-4f7f-b3e8-d37312ef72fe)

我不知道我哪里写错了，谢谢帮助！

**Comment by JOU-amjs** [comment]
你在修改后顺序是正常的，但你在`onSuccess`中返回的是`json.data`，但你这次请求返回的数据是`{code: 1, message: '账号不存在'}`，获取`json.data`自然也就是`undefined`了。

我想再解释下为什么第一种情况会顺序相反，是因为alova的`useRequest`并不返回promise对象，因此代码`await useAlovaFetch(...)`其实比异步响应`onSuccess`先执行，这可能再去弄懂下promise和async/await，推荐你[看这篇文章](https://juejin.cn/post/7263089207128850489)。

最后还有一个小小的建议，是关于use hook使用规范的，use hook一般只会在最外层使用，而不会在语句块中使用比如if、for里或事件函数里，[可以看这边](https://alova.js.org/zh-CN/tutorial/get-started/quick-start#use-hook-%E7%9A%84%E4%BD%BF%E7%94%A8%E8%A7%84%E8%8C%83)，虽然在vue里这样用运行是没问题的。

关于何时使用`useRequest`何时使用method对象发送请求，可以[参考这边](https://alova.js.org/zh-CN/tutorial/best-practice/skills#%E5%8F%91%E9%80%81%E8%AF%B7%E6%B1%82-userequest-or-method)

**Reply by FatNerdPeng** [reply]
噢，谢谢
